### PR TITLE
RSDK-2811 - do not add classify/detect funcs if they are not functional

### PR DESCRIPTION
**Pull request by kim-mishra**

### CODE REVIEW

File: services/vision/mlvision/ml_model.go [hunk 1]
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
 	if err != nil {
 		return nil, err
 	}
+
 	classifierFunc, err := attemptToBuildClassifier(mlm)
 	if err != nil {
-		logger.Infow("error turning turn ml model into a classifier:", "model", params.ModelName, "error", err)
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
```
Comment (bhaney) [inline services/vision/mlvision/ml_model.go:75]: Keep this as Infow in order to allow the logger to format the information as it sees fit

File: services/vision/mlvision/ml_model.go [hunk 2]
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
...
+			logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
+		} else {
+			logger.Infof("model %q fulfills a vision service ciassifier", params.ModelName)
+		}
 	}
+
 	detectorFunc, err := attemptToBuildDetector(mlm)
 	if err != nil {
-		logger.Infow("error turning turn ml model into a detector:", "model", params.ModelName, "error", err)
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
```
Comment (bhaney) [inline services/vision/mlvision/ml_model.go:87]: keeps this as Infow

File: services/vision/mlvision/ml_model.go [hunk 3]
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
...
 	} else {
-		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+		detectorFunc, err = checkIfDetectorWorks(ctx, detectorFunc)
+		if err != nil {
+			logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
```
Comment (bhaney) [inline services/vision/mlvision/ml_model.go:91]: Turn this into Infow

File: services/vision/mlvision/ml_model.go [hunk 4]
```diff
@@ -92,9 +105,12 @@ func registerMLModelVisionService(
 }
 
 // Unpack output based on expected type and force it into a []float64.
-func unpack(inMap map[string]interface{}, name string) []float64 {
+func unpack(inMap map[string]interface{}, name string) ([]float64, error) {
 	var out []float64
 	me := inMap[name]
+	if me == nil {
+		return nil, errors.New("Invalid input to unpack")
```
Comment (bhaney) [inline services/vision/mlvision/ml_model.go:112]: add the name into the error as "no such tensor named %q to unpack"

**Review by bhaney — CHANGES_REQUESTED** [review]
mostly formatting and image related comments

File: services/vision/mlvision/ml_model.go [hunk 5]
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
 	if err != nil {
 		return nil, err
 	}
+
 	classifierFunc, err := attemptToBuildClassifier(mlm)
 	if err != nil {
-		logger.Infow("error turning turn ml model into a classifier:", "model", params.ModelName, "error", err)
+		logger.Infow("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
```
Comment (bhaney) [inline services/vision/mlvision/ml_model.go:75]: The way `Infow` works is like a map: You have an initial message, and then a map of messages, like:
```
log.Infow("here is relevant info",
"timestamp", time.Now(),
"name", thing.Name,
"error", err,
)
```

So in your case, you'll want to write something like
`logger.Infow("was not able to turn ml model into classifier", "ml model", params.ModelName, "error", err)`

File: services/vision/mlvision/ml_model.go [hunk 6]
```diff
@@ -69,18 +69,31 @@ func registerMLModelVisionService(
...
 	} else {
-		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+		classifierFunc, err = checkIfClassifierWorks(ctx, classifierFunc)
+		if err != nil {
+			logger.Infow("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
```
Comment (bhaney) [inline services/vision/mlvision/ml_model.go:79]: same here, rather than using Wrapf, use the dict structure of Infow

File: services/vision/mlvision/classifier.go [hunk 7]
```diff
@@ -45,9 +45,15 @@ func attemptToBuildClassifier(mlm mlmodel.Service) (classification.Classifier, e
...
+		probs, err := unpack(outMap, "probability")
+		if err != nil {
+			return nil, err
+		}
 		if len(probs) == 0 {
-			probs = unpack(outMap, DefaultOutTensorName+"0")
+			probs, err = unpack(outMap, DefaultOutTensorName+"0")
+			if err != nil {
+				return nil, err
+			}
```
Comment (bhaney) [inline services/vision/mlvision/classifier.go:56]: Would you ever get to this statement? If probability was not able to be unpacked, wouldn't it just return the error and not come to this statement?

File: services/vision/mlvision/classifier.go [hunk 8]
```diff
@@ -62,3 +68,17 @@ func attemptToBuildClassifier(mlm mlmodel.Service) (classification.Classifier, e
 		return classifications, nil
 	}, nil
 }
+
+func checkIfClassifierWorks(ctx context.Context, cf classification.Classifier) (classification.Classifier, error) {
+	if cf == nil {
+		return nil, errors.New("Nil classifier function")
+	}
+
+	img := &image.RGBA{}
```
Comment (bhaney) [inline services/vision/mlvision/classifier.go:77]: This initializes as a (0,0) dimensional image. It's quite an edge case and I wouldn't normally expect this to work. Have you tested this in an integration test? You should add tests to see if these functions behave as expected.

File: services/vision/mlvision/detector.go [hunk 9]
```diff
@@ -51,17 +51,35 @@ func attemptToBuildDetector(mlm mlmodel.Service) (objectdetection.Detector, erro
 			return nil, err
 		}
 
-		locations := unpack(outMap, "location")
+		locations, err := unpack(outMap, "location")
+		if err != nil {
+			return nil, err
+		}
 		if len(locations) == 0 {
```
Comment (bhaney) [inline services/vision/mlvision/detector.go:58]: if the original "location" is not unpacked, would you ever get to this line?

File: services/vision/mlvision/detector.go [hunk 10]
```diff
@@ -83,3 +101,17 @@ func attemptToBuildDetector(mlm mlmodel.Service) (objectdetection.Detector, erro
 		return detections, nil
 	}, nil
 }
+
+func checkIfDetectorWorks(ctx context.Context, df objectdetection.Detector) (objectdetection.Detector, error) {
```
Comment (bhaney) [inline services/vision/mlvision/detector.go:105]: Make a test for this to make sure it works as expected

**Review by bhaney — CHANGES_REQUESTED** [review]
some more questions, and Infow suggestions

File: services/vision/mlvision/classifier.go (hunk 8 above)
Comment (kim-mishra) [reply services/vision/mlvision/classifier.go:77]: I tested this with the model that I had been using earlier and it worked, I can write a test as well!

File: services/vision/mlvision/detector.go (hunk 9 above)
Comment (kim-mishra) [reply services/vision/mlvision/detector.go:58]: I believe so because they are using different keys to get data out of `outputMap`

File: services/vision/mlvision/classifier.go (hunk 8 above)
Comment (bhaney) [reply services/vision/mlvision/classifier.go:77]: I meant in that often there is a "resize" operations that gets done on images when they get fed into an ML model to make them comform to what the model expects. "resizing" an (0,0) image seems like it would have caused a problem, but I guess it didn't!
